support customizing JWT claims (close #3485) (#3575)
hasura/graphql-engine: CHANGELOG.md

## Next release

> ### Server - Support for custom JWT Claims

Reviewer: Support for mapping session variables to default JWT claims

- server: some mutations that cannot be performed will no longer be in the schema (for instance, `delete_by_pk` mutations won't be shown to users that do not have select permissions on all primary keys) (#4111)
- server: miscellaneous description changes (#4111)
- server: treat the absence of `backend_only` configuration and `backend_only: false` equally (closing #5059) (#4111)
- server: support customizing JWT claims (close #3485)

> Some providers don't let users add custom JWT claims. In such a case, the server will provide a JWT configuration option to specify a mapping of hasura session variables to values in existing claims via JSONPath/literal values. The JWT config now supports an extra optional field `claims_map` which is an JSON object which maps from session variables to JSON paths.

Reviewer: ```suggestion
Some auth providers do not let users add custom claims in JWT. In such cases, the server can take a JWT configuration option called `claims_map` to specify a mapping of Hasura session variables to values in existing claims via JSONPath/literal values. 
```

Example:-

Consider the following JWT claim:

```
  {
    "sub": "1234567890",
    "name": "John Doe",
    "admin": true,
    "iat": 1516239022,
    "user": {
      "id": "ujdh739kd",
      "appRoles": ["user", "editor"]
    }
  }
```

> The corresponding JWT config should be:

Reviewer: ```suggestion
The corresponding JWT config can be:
```

- server: fix importing of allow list query from metadata (fix #4687)
- server: flush log buffer during shutdown (#4800)
- server: fix edge case with printing logs on startup failure (fix #4772)

> - server: support customizing JWT claims using 'claims_map' in JWT config (fix #3485)

Reviewer 2: Oh sorry, maybe it got added due to a bad merge. I'll remove the second one. 

@tirumaraiselvan what do you think about the comment part?

> - server: support customizing JWT claims using 'claims_map' in JWT config (fix #3485)

Reviewer: The comment part is supposed to become a highlight so you can add it above `Bug fixes and improvement` section. For example, something like this: https://github.com/hasura/graphql-engine/blob/master/CHANGELOG.md#server-debugging-mode-for-non-admin-roles
